src/libstd/path.rs

```diff
@@ -0,0 +1,2415 @@
+        debug_assert!(self.front == State::Body);
+        let (extra, comp) = match self.path.iter().position(|b| self.is_sep(*b)) {
+            None => (0, self.path),
+            Some(i) => (1, &self.path[.. i]),
+        };
+        (comp.len() + extra, parse_single_component(comp))
+    }
+
+    // parse a component from the right, saying how many bytes to consume to
+    // remove the component
+    fn parse_next_component_back(&self) -> (usize, Option<Component<'a>>) {
+        debug_assert!(self.back == State::Body);
+        let start = self.prefix_and_root();
+        let (extra, comp) = match self.path[start..].iter().rposition(|b| self.is_sep(*b)) {
+            None => (0, &self.path[start ..]),
+            Some(i) => (1, &self.path[start + i + 1 ..]),
+        };
+        (comp.len() + extra, parse_single_component(comp))
+    }
+
+    // trim away repeated separators (i.e. emtpy components) on the left
+    fn trim_left(&mut self) {
+        while !self.path.is_empty() {
+            let (size, comp) = self.parse_next_component();
+            if comp.is_some() {
+                return;
+            } else {
+                self.path = &self.path[size ..];
+            }
+        }
+    }
+
+    // trim away repeated separators (i.e. emtpy components) on the right
+    fn trim_right(&mut self) {
+        while self.path.len() > self.prefix_and_root() {
+            let (size, comp) = self.parse_next_component_back();
+            if comp.is_some() {
+                return;
+            } else {
+                self.path = &self.path[.. self.path.len() - size];
+            }
+        }
+    }
+
+    /// Examine the next component without consuming it.
+    pub fn peek(&self) -> Option<Component<'a>> {
+        self.clone().next()
+    }
+}
+
+impl<'a> Iter<'a> {
+    /// Extract a slice corresponding to the portion of the path remaining for iteration.
+    pub fn as_path(&self) -> &'a Path {
+        self.inner.as_path()
+    }
+}
+
+impl<'a> Iterator for Iter<'a> {
+    type Item = &'a OsStr;
+
+    fn next(&mut self) -> Option<&'a OsStr> {
+        self.inner.next().map(Component::as_os_str)
+    }
+}
+
+impl<'a> DoubleEndedIterator for Iter<'a> {
+    fn next_back(&mut self) -> Option<&'a OsStr> {
+        self.inner.next_back().map(Component::as_os_str)
+    }
+}
+
+impl<'a> Iterator for Components<'a> {
+    type Item = Component<'a>;
+
+    fn next(&mut self) -> Option<Component<'a>> {
+        if self.finished() {
+            None
+        } else if self.front == State::Prefix {
+            self.front = State::Root;
+            if self.prefix_len() > 0 {
+                debug_assert!(self.prefix_len() <= self.path.len());
+                let prefix = &self.path[.. self.prefix_len()];
+                self.path = &self.path[self.prefix_len() .. ];
+                Some(Component::Prefix(unsafe { Path::from_u8_slice(prefix) }))
+            } else {
+                self.next()
```

Comment:
This may be better structured as a `loop` + `match` instead of recursively calling `self.next()`